From ce417edad38c72583d8458aacaea6821ea8beb37 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Tue, 12 Jul 2022 19:52:13 +0300
Subject: [PATCH] [SW-95553] gaudi3: disable pif isolation before pcie configuration

By default, the PIF isolation is enabled for slave die, that means
any access to PCIe domain register will result into hang. This patch
add changes to disable PIF isolation if pcie_ctrl fuse value is enabled.

Note that the PIF isolation is by default disabled on master die,
however it disabled for both dies based on die_id for safer side.

Change-Id: Ib57dd15cbbbca3d55325d2f85df873daa4b1fc4f
---

diff --git a/arch/gaudi3/arc/recovery.c b/arch/gaudi3/arc/recovery.c
index 1cd6d84..14815fe 100644
--- a/arch/gaudi3/arc/recovery.c
+++ b/arch/gaudi3/arc/recovery.c
@@ -98,6 +98,18 @@
 	arcboot_rom_status_update_bringup(img_sts_base, PROGRESS_EROM_IMG_FINISH);
 }
 
+/* TODO: remove below macros when added in specs/gaudi3 */
+#define D0_PSOC_BOOT_CONF_D0_PIF_ISO		0xC493750
+#define D0_PSOC_BOOT_CONF_D1_PIF_ISO		0xC493754
+
+static void disable_pif_isolation(void)
+{
+	if (arcboot_rom_global.die_id)
+		regmap_writel(0x0, D0_PSOC_BOOT_CONF_D1_PIF_ISO);
+	else
+		regmap_writel(0x0, D0_PSOC_BOOT_CONF_D0_PIF_ISO);
+}
+
 static void rom_pcie_recovery(void)
 {
 	uint32_t pcie_ctrl_val = arcboot_rom_global.config_fuse & EFUSE_SYSCFG_PCIE_CONTROL_MASK;
@@ -108,6 +120,7 @@
 	 */
 	if ((arcboot_rom_global.die_id && pcie_ctrl_val) ||
 		(!arcboot_rom_global.die_id && !pcie_ctrl_val)) {
+		disable_pif_isolation();
 		/* Disable PCIe error responses (both HBW and LBW) */
 		regmap_writel(0x1, mmPCIE_RSP_OK);
 		regmap_writel(0x1, mmD0_PCIE_WRAP_PCIE_RSP_OK_1);
